lib/view_component/compiler.rb

```diff
@@ -40,28 +40,30 @@ def compile(raise_errors: false)
         component_class.validate_collection_parameter!
       end
 
-      templates.each do |template|
-        # Remove existing compiled template methods,
-        # as Ruby warns when redefining a method.
-        method_name = call_method_name(template[:variant])
+      component_class.lock.synchronize do
```

Comment:
Can we move this to be just under the `return if compiled?` check to cover the entire method? 

Additionally, I think we may want to implement something like `CompilerCache.invalidate_component!(component_klass)` so that any other callers will block on the monitor instead of trying to continue rendering due to the `return if compiled?` call on the first line of this method.

Reply:
Good idea. Will do it!

Reply:
Done :)